### DISCUSSION TITLE
What is the proper way to close result set from server side cursor select that doesn't hang the program?

### DISCUSSION THREAD

**Discussion by michael-newsrx** [sqlalchemy/sqlalchemy#9356] [Usage Questions] [answered]
I'm running a query on millions of records and need to use server side cursors. The DB is MariaDB. I'm able to get streaming working, but when I close the connection either via the context manager or via an explicit #close() everything hangs and pulls in and discards the remaining data associated with the server side cursor. How can I stop the streaming if the code detects a condition which indicates that no more processing on additional rows of data should be done?

Shouldn't the #close() on a result object stop the steaming of data from the server instead of causing a potentially many hour or longer hang waiting on the result set to be streamed over the network then discarded?

Code posted to: https://stackoverflow.com/questions/75538278/sqlalchemy-with-server-side-cursors-and-need-to-close-the-result-set-early-but

**Comment by zzzeek** [comment, marked as answer]
All the MySQL drivers can't exit a server side result set unless all rows are exhausted.  To the degree that the individual drivers perform this for you when their cursor.close() method is called I'm not sure about.    Sqlalchemy however just calls the underlying methods on the cursor, so if you are seeking guidance / grievance here your best bet is to demonstrate the behavior using the driver directly and take it up with the maintainers of that driver, assuming it can be reproduced.


If otoh your observation is that driver used directly behaves differently, please provide MCVE examples here (directly, no stackoverflow links please) of both sqlalchemy and driver code and I can make sure they exhibit equivalent behaviors.

The "real" answer here is not use server side cursor options with MySQL or mariadb, as these features are well known to be immature and extremely error prone.  use result pagination with client side cursors instead (which you probably have to do anyway since your use case seems to be that you only want to read a subset of rows.  The MySQL drivers do not offer an adequate option for this pattern).

**Reply by michael-newsrx** [reply]
OK, thanks, I was finally able to find where the MySQL driver simply exhausts the cursor. The MariaDb driver doesn't have that in the python code, but I suspect something similar is happening in the underlying mariadb-client *.so, which means that if there is any fix, it will have to come from the MariaDB project. The interesting thing is the following MariaDB code which indicates that in theory should be possible:

```python
class cursor(object):
    """ Returns a MariaDB cursor object """
    def close(self): # real signature unknown; restored from __doc__
        """
        Closes the cursor. If the cursor has pending or unread results, .close()
        will cancel them so that further operations using the same connection
        can be executed.
        
        The cursor will be unusable from this point forward; an Error (or subclass)
        exception will be raised if any operation is attempted with the cursor.
        """
        pass
```

**Reply by zzzeek** [reply]
The mariadb driver is not that mature and we've had problems with it.   In production I would only use mysqlclient or pymysql.
